src/Common/src/Interop/Windows/sspicli/SecuritySafeHandles.cs

```diff
@@ -464,141 +429,100 @@ internal abstract partial class SafeDeleteContext : SafeHandle
-                                        ref inCredentials,
-                                        isContextAbsent,
-                                        (byte*)(((object)targetName == (object)dummyStr) ? null : namePtr),
-                                        inFlags,
-                                        endianness,
-                                        haveInSecurityBufferDescriptor ? &inSecurityBufferDescriptor : null,
-                                        refContext,
-                                        ref outSecurityBufferDescriptor,
-                                        ref outFlags,
-                                        outFreeContextBuffer);
-                    }
+                        string punyCode = s_idnMapping.GetAscii(targetName);
+                        fixed (char* namePtr = punyCode)
+                        {
+                            errorCode = MustRunInitializeSecurityContext(
+                                            ref inCredentials,
+                                            isContextAbsent,
+                                            (byte*)(((object)targetName == (object)dummyStr) ? null : namePtr),
+                                            inFlags,
+                                            endianness,
+                                            &inSecurityBufferDescriptor,
+                                            refContext,
+                                            ref outSecurityBufferDescriptor,
+                                            ref outFlags,
+                                            outFreeContextBuffer);
+                        }
 
-                    if (NetEventSource.IsEnabled) NetEventSource.Info(null, "Marshalling OUT buffer");
+                        if (NetEventSource.IsEnabled) NetEventSource.Info(null, "Marshalling OUT buffer");
 
-                    // Get unmanaged buffer with index 0 as the only one passed into PInvoke.
-                    outSecBuffer.size = outUnmanagedBuffer.cbBuffer;
-                    outSecBuffer.type = outUnmanagedBuffer.BufferType;
-                    if (outSecBuffer.size > 0)
-                    {
-                        outSecBuffer.token = new byte[outSecBuffer.size];
-                        Marshal.Copy(outUnmanagedBuffer.pvBuffer, outSecBuffer.token, 0, outSecBuffer.size);
-                    }
-                    else
-                    {
-                        outSecBuffer.token = null;
+                        // Get unmanaged buffer with index 0 as the only one passed into PInvoke.
+                        outSecBuffer.size = outUnmanagedBuffer.cbBuffer;
+                        outSecBuffer.type = outUnmanagedBuffer.BufferType;
+                        outSecBuffer.token = outSecBuffer.size > 0 ?
+                            new Span<byte>((byte*)outUnmanagedBuffer.pvBuffer, outUnmanagedBuffer.cbBuffer).ToArray() :
```

Comment:
Curious, is this a significant source of allocations? Seems like it would be, and we should fix it.

Reply:
It's not insignificant, but it's also not at the top of the list, nor do I currently have a straightforward way to fix it. Suggestions welcome.

Reply:
Do pointer math and return a Span<byte>. The returned pvBuffer/cbBuffer will always be a subset of the originally passed in buffer.

Reply:
> The returned pvBuffer/cbBuffer will always be a subset of the originally passed in buffer.

Hmm, that makes sense but wasn't immediately obvious to me.  I'll take another look.

> Do pointer math

That shouldn't be necessary. If this is actually in one of the managed buffers, which are all pinned, then just constructing the span as I already do will end up being indistinguishable to the runtime from recalculating the offset inside one of the buffers and constructing a span from it, as inside span it's just stored as a ref, regardless of whether the input was a pointer or object.  I'll validate then that the span is entirely part of one of the existing buffers.

Reply:
> Do pointer math

Ah, after looking at it again, I understand the suggestion now... I can just set outSecBuffer.offset rather than setting the token.

Reply:
This is messier than it looks, actually, as the resulting byte[] is then passed around in a variety of ways that aren't friendly to spans, so we'd need to either fix up all of those (which isn't currently feasible due to some of these places storing the byte[] on the heap) or also pass around the offset and length.  I'm going to leave this for later.  At least in my benchmark, this is generating one ~75-byte byte[] for each of the client and server, so while it'd be nice to eliminate, it's not amongst the top concerns right now.